lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt

```diff
@@ -1230,185 +1278,144 @@ internal class EvaluatingCompiler(
+                ErrorCode.SEMANTIC_FLOAT_PRECISION_UNSUPPORTED,
+                errorContextFrom(expr.asType.metas.toPartiQlMetaContainer()),
+                internal = false)
+        }
 
-                    } catch (e: EvaluationException) {
-                        if (e.internal) {
-                            throw e
+        val expThunk = compileAstExpr(expr.value)
+
+        // TODO consider making this more efficient by not directly delegating to CAST
+        // TODO consider also making the operand not double evaluated (e.g. having expThunk memoize)
+        val castThunkEnv = compileCastHelper(expr.value, expr.asType, expr.metas.toPartiQlMetaContainer())
+        return thunkFactory.thunkEnv(metas) { env ->
+            val sourceValue = expThunk(env)
+            try {
+                when {
+                    // NULL/MISSING can cast to anything as themselves
+                    sourceValue.isUnknown() -> valueFactory.newBoolean(true)
+                    else -> {
+                        val castedValue = castThunkEnv(env)
+                        when {
+                            // NULL/MISSING from cast is a permissive way to signal failure
+                            castedValue.isUnknown() -> valueFactory.newBoolean(false)
+                            else -> valueFactory.newBoolean(true)
                         }
-                        valueFactory.newBoolean(false)
                     }
                 }
+
+            } catch (e: EvaluationException) {
+                if (e.internal) {
+                    throw e
+                }
+                valueFactory.newBoolean(false)
             }
-            TypedOp.CAN_LOSSLESS_CAST -> {
-                // TODO consider making this more efficient by not directly delegating to CAST
-                val castThunkEnv = compileCast(typedOpParameter.staticType)
-                thunkFactory.thunkEnv(metas) { env ->
-                    val sourceValue = expThunk(env)
-                    val sourceType = StaticType.fromExprValue(sourceValue)
+        }
+    }
 
-                    fun roundTrip(): ExprValue {
-                        val castedValue = castThunkEnv(env)
+    private fun compileCanLosslessCast(expr: PartiqlAst.Expr.CanLosslessCast, metas: MetaContainer): ThunkEnv {
+        val typedOpParameter = expr.asType.toTypedOpParameter(customTypedOpParameters)
+        if (typedOpParameter.staticType is AnyType) {
+            return thunkFactory.thunkEnv(metas) { valueFactory.newBoolean(true) }
+        }
+        if (compileOptions.typedOpBehavior == TypedOpBehavior.HONOR_PARAMETERS && expr.asType is PartiqlAst.Type.FloatType && expr.asType.precision != null) {
+            err(
+                "FLOAT precision parameter is unsupported",
+                ErrorCode.SEMANTIC_FLOAT_PRECISION_UNSUPPORTED,
+                errorContextFrom(expr.asType.metas.toPartiQlMetaContainer()),
+                internal = false)
+        }
```

Comment:
I think this check can be removed. It's already done by `compileCastHelper` https://github.com/partiql/partiql-lang-kotlin/blob/6ec9e0ad24b29152e94db19df073e981c652059b/lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt#L1205-L1211

Reply:
Will remove it in the next PR.